plugins/clickhouse-schema-management/main.go

```diff
@@ -0,0 +1,338 @@
+	defer clickhouseMigrate.Close()
+	if err := startMigration(clickhouseMigrate); err != nil {
+		klog.ErrorS(err, "Error when migrating")
+	}
+}
+
+func initMigration() (*migrate.Migrate, error) {
+	// Copy migrators from tmp path to persistent path for the downgrading usage in the future
+	if err := copyMigrators(); err != nil {
+		return nil, fmt.Errorf("error when copying migrators: %v", err)
+	}
+	// versionMap is used to map Theia version string to golang-migrate version number
+	if err := initializeVersionMap(); err != nil {
+		return nil, fmt.Errorf("error when generating version number map: %v", err)
+	}
+	userName := getEnv("MIGRATE_USERNAME")
+	password := getEnv("MIGRATE_PASSWORD")
+	databaseURL := getEnv("DB_URL")
+	if len(userName) == 0 || len(password) == 0 || len(databaseURL) == 0 {
+		return nil, fmt.Errorf("unable to load environment variables, MIGRATE_USERNAME, MIGRATE_PASSWORD and DB_URL must be defined")
+	}
+	clickHouseURL = fmt.Sprintf("%s?username=%s&password=%s", databaseURL, userName, password)
+	migrateDatabaseURL := fmt.Sprintf("clickhouse://%s&x-multi-statement=true", clickHouseURL)
+	migrateSourceURL := fmt.Sprintf("file://%s", migratorPersistentPath)
+	clickhouseMigrate, err := newMigrate(migrateSourceURL, migrateDatabaseURL)
+	if err != nil {
+		return nil, fmt.Errorf("error when creating a Migrate instance for ClickHouse: %v", err)
+	}
+	return clickhouseMigrate, nil
+}
+
+// Get Theia version and data version number, migrate if they are different
+func startMigration(clickhouseMigrate *migrate.Migrate) error {
+	theiaVersionNumber, err := getTheiaVerstionNumber()
+	if err != nil {
+		return fmt.Errorf("error when getting Theia version: %v", err)
+	}
+	dataVersionNumber, err := getDataVersionNumber(*clickhouseMigrate)
+	if err != nil {
+		return fmt.Errorf("error when getting the data version: %v", err)
+	}
+	if theiaVersionNumber == dataVersionNumber {
+		klog.InfoS("Data schema version is the same as Theia version. Migration skipped.")
+	} else if dataVersionNumber == -1 {
+		klog.InfoS("No existing data schema. Migration skipped.")
+	} else {
+		klog.InfoS("Migrate data schema", "from", dataVersionNumber, "to", theiaVersionNumber)
+		err = clickhouseMigrate.Steps(theiaVersionNumber - dataVersionNumber)
+		if err != nil {
+			return fmt.Errorf("error when applying migrations: %v", err)
+		}
+	}
+	// Set the data schema version to Theia version anyway, as we expect initial
+	// data will be created even if the migration is skipped.
+	err = clickhouseMigrate.Force(theiaVersionNumber)
+	if err != nil {
+		return fmt.Errorf("error when setting version: %v", err)
+	}
+	return nil
+}
+
+func copyMigrators() error {
+	bashCmd := fmt.Sprintf("mkdir -p %s", migratorPersistentPath)
+	cmd := execCommand("bash", "-c", bashCmd)
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("error when creating folder: %s, error: %v", bashCmd, err)
+	}
+	bashCmd = fmt.Sprintf("cp -r %s/* %s", migratorTmpPath, migratorPersistentPath)
```

Comment:
While I do not love shelling out, I agree that using os.read & os.write would be a bit cumbersome.

My only advice would be - if possible - to not open a shell, and distinguish command "cp" from arguments.
Eg.:

execCommand("cp", fmt.Sprintf(" -r %s/* %s", migratorTmpPath, migratorPersistentPath))

As migratorTmpPath/migratorPersistentPath are constants there's no need to sanitize them.

These measures might not be needed here but it's always good to implement measures to prevent shell injection (some more info here: https://semgrep.dev/docs/cheat-sheets/go-command-injection/)

Reply:
It makes sense for me. Updated, thanks!